File: ExpNameDb/trunk/src/ExpNameDatabase.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class ExpNameDatabase...
//
// Author List:
//      Andy Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "ExpNameDb/ExpNameDatabase.h"

//-----------------
// C/C++ Headers --
//-----------------
#include <fstream>
#include <boost/format.hpp>

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "ExpNameDb/Exceptions.h"
#include "MsgLogger/MsgLogger.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

namespace {

  const char logger[] = "ExpNameDatabase";
}


//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace ExpNameDb {

//----------------
// Constructors --
//----------------
ExpNameDatabase::ExpNameDatabase (const std::string fname)
  : m_path(fname)
{
  if (m_path.path().empty()) {
    MsgLog(logger, error, "Failed to find database file " << fname);
    throw FileNotFoundError(ERR_LOC, fname);
  }
}

/**
 *  @brief Get instrument and experiment name given experiment ID.
 *
 *  @param[in] id    Experiment ID.
 *  @return Pair of strings, first string is instrument name, second is experiment name,
 *         both will be empty if ID is not known.
 *
 */
std::pair<std::string, std::string>
ExpNameDatabase::getNames(unsigned id) const
{
  // open file and read it
  std::ifstream db(m_path.path().c_str());
  unsigned dbExpNum;
  std::string instrName;
  std::string expName;
  std::pair<std::string, std::string> res;
  while (db >> dbExpNum >> instrName >> expName) {
    if (dbExpNum == id) {
      res = std::make_pair(instrName, expName);
      break;
    }
  }
  MsgLog(logger, debug, boost::format("ExpNameDatabase::getNames(%1%) -> (%2%, %3%)") % id % res.first % res.second);
  return res;
}

/**
 *  @brief Get experiment ID given instrument and experiment names.
 *
 *  Instrument name may be empty, if experiment name is unambiguous. If instrument name
 *  is empty and experiment name is ambiguous then first matching ID is returned.
 *
 *  @param[in] instrument   Instrument name.
 *  @param[in] experiment   Experiment name.
 *  @return Experiment ID or 0 if instrument/experiment is not known.
 *
 */
unsigned
ExpNameDatabase::getID(const std::string& instrument, const std::string& experiment) const
{
  // open file and read it
  std::ifstream db(m_path.path().c_str());
  unsigned dbExpNum;
  std::string instrName;
  std::string expName;
  unsigned res = 0;
  while (db >> dbExpNum >> instrName >> expName) {
    if (expName == experiment and (instrument.empty() or instrName == instrument)) {
      res = dbExpNum;
      break;
    }
  }
  MsgLog(logger, debug, boost::format("ExpNameDatabase::getID(%1%, %2%) -> %3%") % instrument % experiment % res);
  return res;
}

/**
 *  @brief Get instrument name and experiment ID for given experiment name.
 *
 *  If experiment name is ambiguous then first matching name and ID is returned.
 *
 *  @param[in] experiment   Experiment name.
 *  @return Pair of instrument name and experiment ID, name will be empty if experiment is not known.
 *
 */
std::pair<std::string, unsigned>
ExpNameDatabase::getInstrumentAndID(const std::string& experiment) const
{
  // open file and read it
  std::ifstream db(m_path.path().c_str());
  unsigned dbExpNum;
  std::string instrName;
  std::string expName;
  std::pair<std::string, unsigned> res;
  while (db >> dbExpNum >> instrName >> expName) {
    if (expName == experiment) {
      res = std::make_pair(instrName, dbExpNum);
    }
  }
  MsgLog(logger, debug, boost::format("ExpNameDatabase::getInstrumentAndID(%1%) -> (%2%, %3%)") % experiment % res.first % res.second);
  return res;
}


} // namespace ExpNameDb

File: ExpNameDb/trunk/src/ExpNameDatabase.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id$
#
# Description:
#  Module ExpNameDatabase...
#
#------------------------------------------------------------------------

"""Module for a class which provides mapping between experiment names and IDs.

@version $Id$

@author Andy Salnikov
"""


#------------------------------
#  Module's version from CVS --
#------------------------------
__version__ = "$Revision$"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#---------------------------------
#  Imports of base class module --
#---------------------------------

#-----------------------------
# Imports for other modules --
#-----------------------------
from AppUtils.AppDataPath import AppDataPath

#----------------------------------
# Local non-exported definitions --
#----------------------------------

def _tripleGen(fname):
    """generator for triples (expnum, instr, exper)"""
    for line in open(fname):
        num, instr, exper = line.split()
        num = int(num)
        yield num, instr, exper

#------------------------
# Exported definitions --
#------------------------

#---------------------
#  Class definition --
#---------------------
class ExpNameDatabase ( object ) :
    """Class which provides mapping between experiment names and IDs."""

    #----------------
    #  Constructor --
    #----------------
    def __init__ ( self, dbname = "ExpNameDb/experiment-db.dat" ) :
        """
        Constructor takes the name of the file containing the database
        File name is relative with respect to the $SIT_DATA (one of its components).
        If the file is not found an exception is generated.
        """

        # find experiment database file
        self.expdbpath = AppDataPath(dbname).path()
        if not self.expdbpath: raise ValueError("ExpNameDatabase: file name not found: " + dbname)

    #-------------------
    #  Public methods --
    #-------------------

    def getNames(self, id):
        """
        self.getNames(id: int) -> tuple of 2 strings
        
        Get instrument and experiment name given experiment ID. Takes experiment ID and 
        returns pair of strings, first string is instrument name, second is experiment name,
        both will be empty if ID is not known.
        """

        # scan database
        res = ("", "")
        for expnum, instr, exper in _tripleGen(self.expdbpath):
            if expnum == id:
                res = (instr, exper)
                break
        logging.debug("ExpNameDatabase.getNames(%d) -> %s", id, res)
        return res


    def getID(self, instrument, experiment):
        """
        self.getID(instrument: str, experiment: str) -> int
        
        Get experiment ID given instrument and experiment names. Instrument name may be empty 
        if experiment name is unambiguous. If instrument name is empty and experiment name is 
        ambiguous then first matching ID is returned. Returns experiment ID or 0 if 
        instrument/experiment is not known.
        """

        # scan database
        res = 0
        for expnum, instr, exper in _tripleGen(self.expdbpath):
            if exper == experiment and (not instrument or instr == instrument):
                res = expnum
                break
        logging.debug("ExpNameDatabase.getID('%s', '%s') -> %d", instrument, experiment, res)
        return res


    def getInstrumentAndID(self, experiment):
        """
        self.getInstrumentAndID(experiment: str) -> tuple of string and int
        
        Get instrument name and experiment ID for given experiment name. If experiment name is 
        ambiguous then first matching name and ID is returned. Takes experiment name and returns
        pair of instrument name and experiment ID, name will be empty if experiment is not known.
        """

        # scan database
        res = ("", 0)
        for expnum, instr, exper in _tripleGen(self.expdbpath):
            if exper == experiment:
                res = instr, expnum
        logging.debug("ExpNameDatabase.getID('%s') -> %d", experiment, res)
        return res


#
#  In case someone decides to run this module
#
if __name__ == "__main__" :

    # In principle we can try to run test suite for this module,
    # have to think about it later. Right now just abort.
    sys.exit ( "Module is not supposed to be run as main module" )
